### DISCUSSION TITLE
Homebrew install error : "Bad CPU type in executable" - Monterey 12.4

### DISCUSSION THREAD

**Discussion by Tie3** [Homebrew/discussions#3385] [Getting started] [answered]
I have :
macOS Monterey (12.4)
Macbook air (M1)

This is the error when I try to install Homebrew on macOS Terminal (zsh) 

>==> This script will install:
/opt/homebrew/bin/brew
/opt/homebrew/share/doc/homebrew
/opt/homebrew/share/man/man1/brew.1
/opt/homebrew/share/zsh/site-functions/_brew
/opt/homebrew/etc/bash_completion.d/brew
/opt/homebrew

>Press RETURN/ENTER to continue or any other key to abort:
==> /usr/bin/sudo /usr/sbin/chown -R cd:admin /opt/homebrew
==> Downloading and installing Homebrew...

>/bin/bash: line 226: /usr/local/bin/git: Bad CPU type in executable
Failed during: git init -q

Any suggestions on how to fix this ? Thanks.

**Comment by carlocab** [comment, marked as answer]
Where did `/usr/local/bin/git` come from? Is that from an old installation of Homebrew that you migrated from an Intel Mac?

**Comment by Tie3** [comment]
My trainer gave me this link : [https://sourceforge.net/projects/git-osx-installer/](url)

git version 
>2.32.1 (Apple Git-133)

**Reply by carlocab** [reply]
That doesn't work on M1 macOS. You need to uninstall it first, since it's broken. Check the page of the project you linked for uninstallation instructions.

**Reply by Tie3** [reply]
after uninstall /usr/local/bin/git,
Homebrew installed successfully.

Thank you !!

**Comment by zhanwenchen** [comment]
For me, my `git` was installed by homebrew itself, so I had first run the [official homebrew uninstallation script](https://github.com/homebrew/install#uninstall-homebrew) first:

```shell
/bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/HEAD/uninstall.sh)"
```

And then I was able to successfully run the official installation command.

**Reply by Knogobert** [reply]
Saviour!

**Comment by sunil3gs98** [comment]
I have macOS Ventura (13.2.1) MacBook Pro M2 trying to install Homebrew and I am getting error like this:
% /bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/HEAD/install.sh)"

==> This script will install:
/opt/homebrew/bin/brew
/opt/homebrew/share/doc/homebrew
/opt/homebrew/share/man/man1/brew.1
/opt/homebrew/share/zsh/site-functions/_brew
/opt/homebrew/etc/bash_completion.d/brew
/opt/homebrew

Press RETURN/ENTER to continue or any other key to abort:
==> /usr/bin/sudo /usr/sbin/chown -R sunil3gs98:admin /opt/homebrew
==> Downloading and installing Homebrew...
/bin/bash: line 234: /usr/local/bin/git: Bad CPU type in executable
Failed during: /usr/local/bin/git -c init.defaultBranch=master init --quiet

please give a suggestion for this

**Reply by gromgit** [reply]
>```
> /bin/bash: line 234: /usr/local/bin/git: Bad CPU type in executable
>```

What's the output of:
```
ls -l /usr/local/bin/git
file -L /usr/local/bin/git
which -a git
```
?

**Reply by sunil3gs98** [reply]
ls -l /usr/local/bin/git
lrwxr-xr-x  1 root  wheel  14 Feb 27 19:37 /usr/local/bin/git -> ../git/bin/git

  file -L /usr/local/bin/git
/usr/local/bin/git: Mach-O universal binary with 2 architectures: [x86_64:Mach-O 64-bit executable x86_64] [i386]
/usr/local/bin/git (for architecture x86_64):	Mach-O 64-bit executable x86_64
/usr/local/bin/git (for architecture i386):	Mach-O executable i386

which -a git
/usr/local/bin/git
/usr/bin/git

**Reply by gromgit** [reply]
>```
>file -L /usr/local/bin/git
>/usr/local/bin/git: Mach-O universal binary with 2 architectures: [x86_64:Mach-O 64-bit executable x86_64] [i386]
>/usr/local/bin/git (for architecture x86_64): Mach-O 64-bit executable x86_64
>/usr/local/bin/git (for architecture i386): Mach-O executable i386
>```

There's your problem: you're trying to run an Intel `git` binary on an M1 Mac. The fact that it still contains 32-bit code makes me think it's _really_ old.

Easiest solution: `rm -f /usr/local/bin/git`, then run the installer again.

**Reply by sunil3gs98** [reply]
which installer I need to run again can you be more specific on that ? and tell me the command what it is?

**Reply by gromgit** [reply]
The one you were trying to run, that started this entire thread:
>```
> /bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/HEAD/install.sh)"
>```

**Reply by sunil3gs98** [reply]
Thank you so much for your suggestions and it worked and I was able to install Homebrew😊

**Reply by jmrod1003** [reply]
> Easiest solution: `rm -f /usr/local/bin/git`, then run the installer again

I have the same problem, when I run this my terminal says:

rm: /usr/local/bin/git: Permission denied

What can I do to fix this?

**Reply by sunil3gs98** [reply]
sudo rm -f /usr/local/bin/git 
Give this commad it worked for me

**Reply by jmrod1003** [reply]
thank you!

**Comment by zhanwenchen** [comment]
Maybe you need to reinstall git somehow

**Reply by sunil3gs98** [reply]
how can I reinstall git can you please suggest me because I am new to macOS

**Comment by VI7AY** [comment]
sudo rm -f /usr/local/bin/git - This worked for me.

**Comment by nobady90** [comment]
If someone, as in my case, had installed homebrew on the device with an Apple chip, and then switched to an Intel chip, this removes it by searching for the correct paths: [https://raw.githubusercontent.com/nobady90/homebrew_unistall_from_chip_apple/main/uninstall.sh](https://raw.githubusercontent.com/nobady90/homebrew_unistall_from_chip_apple/main/uninstall.sh)

**Reply by SMillerDev** [reply]
Or run the [official uninstaller](https://github.com/Homebrew/install) in Rosetta with `arch -x86_64 <command>`

**Reply by nobady90** [reply]
perfect, I always notice after simple roads ;-)

**Comment by aaronjwood** [comment]
I have a bit of a different scenario that I thought I'd share with others: I do have Rosetta installed and I get the `Bad CPU type in executable` error for `uname` installed via `coreutils` from brew, but only when running under `arch -x86_64`.
```
# My system version works fine.
arch -x86_64 /usr/bin/uname -m
x86_64

file /usr/bin/uname
/usr/bin/uname: Mach-O universal binary with 2 architectures: [x86_64:Mach-O 64-bit executable x86_64] [arm64e:Mach-O 64-bit executable arm64e]
/usr/bin/uname (for architecture x86_64):	Mach-O 64-bit executable x86_64
/usr/bin/uname (for architecture arm64e):	Mach-O 64-bit executable arm64e

arch -x86_64 /opt/homebrew/opt/coreutils/libexec/gnubin/uname
arch: posix_spawnp: /opt/homebrew/opt/coreutils/libexec/gnubin/uname: Bad CPU type in executable

# Brew's version is very different.
file /opt/homebrew/opt/coreutils/libexec/gnubin/uname
/opt/homebrew/opt/coreutils/libexec/gnubin/uname: Mach-O 64-bit executable arm64
```

I've tried this out with a few other bins, plus some bins I've built myself from our product's source code and the behavior is the same. It seems you can only do this with universal binaries...?

**Reply by SMillerDev** [reply]
On Rosetta you need an x86 slice in the binary. And since brew only compiles native binaries, the arm install in `/opt/homebrew` will never have those.
If you need brew under Rosetta, install it through Rosetta in `/usr/local`.

---

I'd strongly recommend moving away from software that needs it though, because that means that software will stop working with macOS 28: https://developer.apple.com/documentation/apple-silicon/about-the-rosetta-translation-environment
